ldcbench.data-generator/src/main/java/org/dice_research/ldcbench/benchmark/DataGenerator.java

```diff
@@ -225,6 +234,47 @@ protected void addInterlinks(GraphBuilder g) {
         }
     }
 
+    /**
+     * Append a given number of Blank Nodes to an existing graph.
+     * Edges are created to link to the Blank Nodes.
+     *
+     * @param g the graph
+     * @param nodeCount
+     *            the number of blank nodes
+     * @param seed
+     */
+    private void addBlankNodes(GraphBuilder g, int nodeCount, long seed) {
+        int blankNodesIndex = g.getNumberOfNodes();
+        g.setBlankNodesRange(blankNodesIndex, nodeCount);
+        addBNodesOrLiterals(g, blankNodesIndex, nodeCount, seed);
+    }
+
+    /**
+     * Append a given number of literals to an existing graph.
+     * Edges are created to link to the literals.
+     *
+     * @param g the graph
+     * @param nodeCount
+     *            the number of literals
+     * @param seed
+     */
+    private void addLiterals(GraphBuilder g, int nodeCount, long seed) {
+        int literalIndex = g.getNumberOfNodes();
+        g.setLiteralsRange(literalIndex, nodeCount);
+        addBNodesOrLiterals(g, literalIndex, nodeCount, seed);
+    }
+
+    private void addBNodesOrLiterals(GraphBuilder g, int index, int NumberOfNodes, long seed) {
+        Random generator = new Random(seed);
+        g.addNodes(NumberOfNodes);
+        for (int i = index; i < g.getNumberOfNodes(); i++) {
+            //Get a random Source Node
+            int sourceNode = generator.nextInt(index);
```

Comment:
I think this loop is not correct. It should be something like 
```java
        int numberOfIriNodes = ... // get the number of IRI nodes in the graph (i.e., nodes that are neither blank nodes nor literals)
        for (int i = 0; i < numberOfNodes; i++) {
            //Get a random Source Node
            int sourceNode = generator.nextInt(numberOfIriNodes);
```
where `numberOfNodes` is the number of blank nodes / literals that should be added to the graph.

With your current loop, you could use blank nodes or literals as subjects, which can not be handled by other parts of the code.

Reply:
This problem is still not fixed.

Reply:
Could you please double check this ? i think it works because in the loop we then create the edge with 
`g.addEdge(sourceNode, i, propertyId);` at line 274. With **i** been the id of the target so the blankNode or literal. This is why i start from index ( so the id  of the first blankNode/ Literal). At line 269, the BlankNodes/Literals are added to the graph so `g.getNumberOfNodes()` returns the index of the last node. So i think we are in the right intervall.

Starting from 0 would work only with further calculations to get the right indexes for the blankNode or Literal i think.

Reply:
Well, it will work for the first iteration without any issues. However, the problem comes in the later iterations. Let's assume the graph has 100 IRIs and you want to add 20 literals. The line `int sourceNode = generator.nextInt(index);` will randomly sample an integer which is smaller than `index`. In the example, it means that for the last literal that will be added (the one with id 119), the code samples from the range [0,118]. However, it should sample only from the IRI part of the IDs (i.e., [0,99]).

Reply:
it's strange because the line you are mentioning now reads `int sourceNode = generator.nextInt(numberOfIriNodes);` as i introduced the variable `numberOfIriNodes` few commits earlier to solve this problem. When i look at the file on the branch, it has this change. But somehow the PR just shows that the file is outdated but does not show the latest version of it.